src/resolvers/shipment/index.ts

```diff
@@ -0,0 +1,82 @@
+import { UserInputError, ApolloError } from 'apollo-server'
+
+import {
+  QueryResolvers,
+  MutationResolvers,
+  ShipmentResolvers,
+  ShipmentInput,
+} from '../../server-internal-types'
+import { sequelize } from '../../sequelize'
+import Shipment from '../../models/shipment'
+import Group from '../../models/group'
+
+const shipmentRepository = sequelize.getRepository(Shipment)
+const groupRepository = sequelize.getRepository(Group)
+
+// Shipment query resolvers
+const listShipments: QueryResolvers['listShipments'] = async () => {
+  return shipmentRepository.findAll()
+}
+
+// Shipment mutation resolvers
+const addShipment: MutationResolvers['addShipment'] = async (
+  _parent,
+  { input },
+  _context,
+) => {
+  if (
+    !input.shippingRoute ||
+    !input.labelYear ||
+    !input.labelMonth ||
+    !input.sendingHubId ||
+    !input.receivingHubId ||
+    !input.status
+  ) {
+    throw new UserInputError('Shipment arguments invalid', {
+      invalidArgs: Object.keys(input).filter(
+        (key) => !input[key as keyof ShipmentInput],
+      ),
+    })
+  }
+
+  const sendingHub = await groupRepository.findByPk(input.sendingHubId)
+  if (!sendingHub) {
+    throw new ApolloError('Sending hub not found')
+  }
+
+  const receivingHub = await groupRepository.findByPk(input.sendingHubId)
+  if (!receivingHub) {
+    throw new ApolloError('Receiving hub not found')
+  }
+
+  return shipmentRepository.create({
+    shippingRoute: input.shippingRoute,
+    labelYear: input.labelYear,
+    labelMonth: input.labelMonth,
+    sendingHubId: input.sendingHubId,
+    receivingHubId: input.receivingHubId,
+    status: input.status,
+    statusChangeTime: new Date(),
+  })
+}
+
+// Shipment custom resolvers
+const sendingHub: ShipmentResolvers['sendingHub'] = async (parent) => {
+  const sendingHub = await groupRepository.findByPk(parent.sendingHubId)
+  if (!sendingHub) {
+    throw new ApolloError('Sending hub not found')
+  }
+
+  return sendingHub
+}
+
+const receivingHub: ShipmentResolvers['receivingHub'] = async (parent) => {
+  const receivingHub = await groupRepository.findByPk(parent.receivingHubId)
+  if (!receivingHub) {
+    throw new ApolloError('Receiving hub not found')
+  }
+
+  return receivingHub
+}
```

Comment:
thought: aaaaaaaah so this is how graphql resolves associations. I've been wondering what this was gonna like look. Thanks!

Reply:
I was wondering the same thing!